django/core/handlers/asgi.py

```diff
@@ -177,15 +178,50 @@ async def handle(self, scope, receive, send):
             body_file.close()
             await self.send_response(error_response, send)
             return
-        # Get the response, using the async mode of BaseHandler.
+        # Try to catch a disconnect while getting response
+        tasks = [
+            asyncio.create_task(self.run_get_response(request)),
+            asyncio.create_task(self.listen_for_disconnect(receive)),
+        ]
+        done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
+        done, pending = done.pop(), pending.pop()
+        # Allow views to handle cancellation.
+        pending.cancel()
+        try:
+            await pending
+        except asyncio.CancelledError:
+            # Task re-raised the CancelledError as expected.
+            pass
+        try:
+            response = done.result()
+        except RequestAborted:
+            body_file.close()
+            return
+        except AssertionError:
+            body_file.close()
+            raise
+        # Send the response.
+        await self.send_response(response, send)
+
+    async def listen_for_disconnect(self, receive):
+        """Listens for disconnect from the client."""
+        message = await receive()
+        if message:
+            assert (
+                message["type"] == "http.disconnect"
+            ), f"Invalid ASGI message after request body: {message['type']}."
```

Comment:
On the assertion... I think it's a genuine **should never happen** case, and it should be (is) re-raised so it errors out, never to be caught. (It would be a mistake in the protocol server implementation)